Author: Daniel Baumann <daniel@debian.org>
Description: Correcting shebang.

diff -Naurp openerp-server.orig/bin/addons/document/ftpserver/ftpserver.py openerp-server/bin/addons/document/ftpserver/ftpserver.py
--- openerp-server.orig/bin/addons/document/ftpserver/ftpserver.py	2008-12-16 11:06:22.000000000 +0000
+++ openerp-server/bin/addons/document/ftpserver/ftpserver.py	2008-12-16 11:28:19.000000000 +0000
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/usr/bin/python
 # ftpserver.py
 #
 #  pyftpdlib is released under the MIT license, reproduced below:
