Author: Daniel Baumann <daniel@debian.org>
Description: Correcting shebang.

diff -Naurp openerp-server.orig/bin/addons/document/ftpserver/ftpserver.py openerp-server/bin/addons/document/ftpserver/ftpserver.py
--- openerp-server.orig/bin/addons/document/ftpserver/ftpserver.py	2009-05-30 10:14:37.000000000 +0000
+++ openerp-server/bin/addons/document/ftpserver/ftpserver.py	2009-05-30 11:02:00.000000000 +0000
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/usr/bin/python
 # -*- encoding: utf-8 -*-
 # ftpserver.py
 #
